be/src/storage/tablet_updates.cpp

```diff
@@ -827,8 +835,61 @@ void TabletUpdates::_apply_rowset_commit(const EditVersionInfo& version_info) {
     auto& upserts = state.upserts();
     for (uint32_t i = 0; i < upserts.size(); i++) {
         if (upserts[i] != nullptr) {
-            index.upsert(rowset_id + i, 0, *upserts[i], &new_deletes);
-            manager->index_cache().update_object_size(index_entry, index.memory_usage());
+            if (conditional_column >= 0) {
+                auto tablet_column = _tablet.tablet_schema().column(conditional_column);
+                std::vector<uint32_t> read_column_ids;
+                read_column_ids.push_back(conditional_column);
+
+                std::vector<uint64_t> old_rowids(upserts[i]->size());
+                index.get(*upserts[i], &old_rowids);
+                std::map<uint32_t, std::vector<uint32_t>> old_rowids_by_rssid;
+                size_t num_default = 0;
+                vector<uint32_t> idxes;
+                RowsetUpdateState::plan_read_by_rssid(old_rowids, &num_default, &old_rowids_by_rssid, &idxes);
+                std::vector<std::unique_ptr<vectorized::Column>> old_columns(1);
+                auto old_column = ChunkHelper::column_from_field_type(tablet_column.type(), tablet_column.is_nullable());
+                old_columns[0] = old_column->clone_empty();
+                get_column_values(read_column_ids, false, old_rowids_by_rssid, &old_columns);
```

Comment:
After call this function, the data order of `old_columns` is not match with `keys`, i think you need to reorder the data according to `idxes`.